### PR TITLE
Integrate TensorRT EP libs into existing GPU Nuget Package (Approach#2)

### PR DESCRIPTION
**Pull request by chilo-ms**
Description: Integrate TensorRT EP shared libraries into existing GPU Nuget Package as part of the goal to create combined gpu package.

We plan to do it with two approaches:

1. Create another Pipeline’s job to separately build TRT EP. And then extract TRT EP shared library and put it into GPU nuget package.
2. Add additional option --use_tensorrt in current build command to include building TRT EP library.

This PR is approach # 2
Once #8687 is done, we can later adopt this approach.

### CODE REVIEW

**Comment by chilo-ms** [comment]
I've tested the newly added code on [TensorRT Nuget package CI](https://aiinfra.visualstudio.com/Lotus/_build/results?buildId=164605&view=results) and all of the window build encountered following warning treaded as an error:
[error]onnxruntime\core\platform\windows\telemetry.cc(50,1): Error C2220: the following warning is treated as an error

@snnn do you know how to fix this? Thanks.

**Comment by jywu-msft** [comment]
> I've tested the newly added code on [TensorRT Nuget package CI](https://aiinfra.visualstudio.com/Lotus/_build/results?buildId=164605&view=results) and all of the window builds encountered following warning treaded as an error:
> [error]onnxruntime\core\platform\windows\telemetry.cc(50,1): Error C2220: the following warning is treated as an error
> 
> @snnn do you know how to fix this? Thanks.

not sure about that error. but separate question, shouldn't we use the existing GPU Nuget Package pipeline instead of creating a new one?

**Review by snnn — CHANGES_REQUESTED** [review]
Before put it into the official packaging pipeline, some code need be cleaned up.

**Comment by snnn** [comment]
Generally speaking, TRT as an EP is an independent component. By adding "--use_tensorrt" to the build command shouldn't affect how the part of onnx runtime.  Please read the code in onnxruntime_providers.cmake from line 480 to line 517.  Most of setting are global.  Like, we don't want to disable warning 4267, 4244, 4996, 4456 globally.  Please do it only for TRT. 

And, there are many compliance related thing you need to do.  Like, BinSkim, code sign, static code analysis.

**Comment by snnn** [comment]
And please review the use of "USE_TENSORRT" C/C++ macro.

File: tools/ci_build/github/azure-pipelines/nuget/templates/test_win.yml [hunk 1]
```diff
@@ -11,12 +11,15 @@ jobs:
...
-  - name: runCodesignValidationInjection
-    value: false
+    OnnxRuntimeBuildDirectory: '$(Build.BinariesDirectory)'
+    SKIPNONPACKAGETESTS: 'ON'
+    runCodesignValidationInjection: false
+    NugetArtifact: $(Build.BinariesDirectory)\nuget-artifact
+    RunTestScript: test\Microsoft.ML.OnnxRuntime.EndToEndTests\runtest.bat
+    DotNetCoreX64Args: ${{ variables.NugetArtifact }} netcoreapp2.1 x64 $(NuGetPackageVersionNumber)
+    DotNetCoreX86Args: ${{ variables.NugetArtifact }} netcoreapp2.1 x86 $(NuGetPackageVersionNumber)
+    DotNetFrameworkX64Args: ${{ variables.NugetArtifact }} net461 x64 $(NuGetPackageVersionNumber)
```
Comment (snnn) [inline tools/ci_build/github/azure-pipelines/nuget/templates/test_win.yml:21]: Why do we need to support net461?

**Comment by chilo-ms** [comment]
> > I've tested the newly added code on [TensorRT Nuget package CI](https://aiinfra.visualstudio.com/Lotus/_build/results?buildId=164605&view=results) and all of the window builds encountered following warning treaded as an error:
> > [error]onnxruntime\core\platform\windows\telemetry.cc(50,1): Error C2220: the following warning is treated as an error
> > @snnn do you know how to fix this? Thanks.
> 
> not sure about that error. but separate question, shouldn't we use the existing GPU Nuget Package pipeline instead of creating a new one?

Just wanted to test on other CI before integrate into existing GPU Nuget Package.

File: tools/ci_build/github/azure-pipelines/nuget/templates/test_win.yml (hunk 1 above)
Comment (chilo-ms) [reply tools/ci_build/github/azure-pipelines/nuget/templates/test_win.yml:21]: Previous code has support for net416 as well
https://github.com/microsoft/onnxruntime/blob/master/tools/ci_build/github/azure-pipelines/nuget/templates/test_win.yml#L87

**Comment by chilo-ms** [comment]
> Generally speaking, TRT as an EP is an independent component. By adding "--use_tensorrt" to the build command shouldn't affect how the other parts of onnx runtime work. Please read the code in onnxruntime_providers.cmake from line 480 to line 517. Most settings are global. Like, you disable warning 4267, 4244, 4996, 4456 globally for all the code. Please do it only for TRT. And warning 4244 should not be disabled. It should be addressed.
> 
> And, there are many compliance related thing you need to do. Like, BinSkim, code sign, static code analysis.

Yes, I will make sure we only disable warning 4267, 4244, 4996, 4456 for TRT build.

But what confuses me is that [previous CI run](https://aiinfra.visualstudio.com/Lotus/_build/results?buildId=163314&view=logs&j=e0d6d385-aade-5fdc-3d57-e4653b60e3cf&t=4d527569-14d1-52de-b737-b6d4d212444a) was successful for TRT build and didn't have this 
"[error]onnxruntime\core\platform\windows\telemetry.cc(50,1): Error C2220: the following warning is treated as an error".
The build commands are the same for previous CI run and current failed run:
- [previous run](https://github.com/microsoft/onnxruntime/blob/512cf179d256216d3f55ed883f355014a1416e2d/tools/ci_build/github/azure-pipelines/nuget/templates/gpu-tensorrt.yml#L11)

- [current run](https://github.com/microsoft/onnxruntime/blob/2c20123904b3bcb5214e77ec96e80fcb4a092f0c/tools/ci_build/github/azure-pipelines/nuget/templates/gpu.yml#L11)

**Comment by jywu-msft** [comment]
> > > I've tested the newly added code on [TensorRT Nuget package CI](https://aiinfra.visualstudio.com/Lotus/_build/results?buildId=164605&view=results) and all of the window builds encountered following warning treaded as an error:
> > > [error]onnxruntime\core\platform\windows\telemetry.cc(50,1): Error C2220: the following warning is treated as an error
> > > @snnn do you know how to fix this? Thanks.
> > 
> > 
> > not sure about that error. but separate question, shouldn't we use the existing GPU Nuget Package pipeline instead of creating a new one?
> 
> Just want to test on other CI before integrate into existing GPU Nuget Package.

I suggest you directly work on the existing GPU Nuget Pipeline (that is the end objective of this PR). otherwise you risk debugging issues with a new pipeline that aren't related.

**Comment by chilo-ms** [comment]
> > > > I've tested the newly added code on [TensorRT Nuget package CI](https://aiinfra.visualstudio.com/Lotus/_build/results?buildId=164605&view=results) and all of the window builds encountered following warning treaded as an error:
> > > > [error]onnxruntime\core\platform\windows\telemetry.cc(50,1): Error C2220: the following warning is treated as an error
> > > > @snnn do you know how to fix this? Thanks.
> > > 
> > > 
> > > not sure about that error. but separate question, shouldn't we use the existing GPU Nuget Package pipeline instead of creating a new one?
> > 
> > 
> > Just want to test on other CI before integrate into existing GPU Nuget Package.
> 
> I suggest you directly work on the existing GPU Nuget Pipeline (that is the end objective of this PR). otherwise you risk debugging issues with a new pipeline that aren't related.

Yes, I switched to debug on the existing GPU Nuget Pipeline.
https://aiinfra.visualstudio.com/Lotus/_build/results?buildId=164699&view=logs&j=2e397d41-23cf-56ee-ebe2-59fdb571d58c
and the Error C2220 didn't show up on this pipeline.

**Comment by chilo-ms** [comment]
Regarding compliances, BinSkim, code sign, static code analysis.

- **BinSkim**: previous GPU nuget pipeline sets DoCompliance: 'false' in gpu.yml, so it seems compliance.yml (contains BinSkim script) doesn't be applied. 
- **code sign**: I ran the GPU Nuget Package Pipeline and found out many [code sign warnings](https://aiinfra.visualstudio.com/Lotus/_build/results?buildId=164699&view=logs&j=be75321b-f218-52cd-aed6-a1b95c9b2977&t=01fbb7de-d587-50a6-e04d-af5cb20fc7bd&l=119), any suggestions?
- static code analysis: Is there any reference here?

**Comment by chilo-ms** [comment]
> And please review the use of "USE_TENSORRT" C/C++ macro.

Did you refer to using this macro in dotnet test command in runtest.bat ?
`%dn% test test\Microsoft.ML.OnnxRuntime.EndToEndTests\Microsoft.ML.OnnxRuntime.EndToEndTests.csproj --no-restore /p:DefineConstants="USE_TENSORRT"` 

If I didn't explicit add this options /p:DefineConstants="USE_TENSORRT", two TensorRT EP related unittests won't be built and run. As you can see [this pipeline run](https://aiinfra.visualstudio.com/Lotus/_build/results?buildId=163134&view=logs&j=a232d62e-9070-532d-3df9-33d39cb8077d&t=976ca948-e541-5e10-e21a-a03a8c272b6e&l=198).

After explicit add this options /p:DefineConstants="USE_TENSORRT", two additional unittests will be built and run.
Please see [here](https://aiinfra.visualstudio.com/Lotus/_build/results?buildId=164715&view=logs&j=a232d62e-9070-532d-3df9-33d39cb8077d&t=976ca948-e541-5e10-e21a-a03a8c272b6e&l=194).

File: tools/ci_build/github/azure-pipelines/nuget/templates/test_win.yml (hunk 1 above)
Comment (snnn) [reply tools/ci_build/github/azure-pipelines/nuget/templates/test_win.yml:21]: Thank you

**Comment by snnn** [comment]
```cmake
    set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} /wd4996 /wd4244 /wd4267 /wd4099 /wd4551 /wd4505 /wd4515 /wd4706 /wd4456 /wd4324 /wd4701 /wd4804 /wd4702 /wd4458 /wd4703")
```

This is global.

**Comment by snnn** [comment]
> > And please review the use of "USE_TENSORRT" C/C++ macro.
> 
> Did you refer to using this macro in dotnet test command in runtest.bat ?

No, I mean the macro in C/C++ code. When you add trt to pipelines, it means you turned the macro from OFF to ON for the existing packaging pipelines. So, what kind of behavior change it does? Generally speaking, we should reduce the use and avoid using it as possible as we can.   For example, the one in https://github.com/microsoft/onnxruntime/blob/91075255a768814e0524db196ecc33b5c4e8fb2b/include/onnxruntime/core/common/string_helper.h is definitely not needed.  And the things in unit tests should be reviewed and removed.  For example, in onnxruntime/test/providers/cpu/reduction/reduction_ops_test.cc, when USE_TENSORRT is defined then a test will be disabled. It is ok if you disable it just for TRT EP, but you shouldn't do that for all the EPs.  We still want this test.

**Comment by snnn** [comment]
What if we remove the #ifdef USE_TENSORRT at https://github.com/microsoft/onnxruntime/blob/master/onnxruntime/core/framework/provider_bridge_ort.cc#L1251 ? So that the CreateTensorRTProviderOptions function is implemented in all the builds whether USE_TENSORRT is ON or OFF? It shouldn't hurt.  If we have two onnxruntime.dll, one  build with TRT EP, another without TRT EP. These two onnxruntime.dll should be the same. Because TRT EP is external. The same thing applies to CUDA.

**Comment by snnn** [comment]
> Regarding compliances, BinSkim, code sign, static code analysis. I got some questions here.
> 
> * **BinSkim**: previous GPU Nuget Package Pipeline sets DoCompliance: 'false' in gpu.yml, so it seems compliance.yml (contains BinSkim script) doesn't be applied.

Nice finding. Let me fix this.


> * **code sign**: I ran the GPU Nuget Package Pipeline and found out many [code sign warnings](https://aiinfra.visualstudio.com/Lotus/_build/results?buildId=164699&view=logs&j=be75321b-f218-52cd-aed6-a1b95c9b2977&t=01fbb7de-d587-50a6-e04d-af5cb20fc7bd&l=119), any suggestions?

Sure. Ping me on Teams next week. Let me explain it to you. 

> * **static code analysis**: Is there any reference here?

Same as the above. Sorry I can't say too much here.

**Comment by jywu-msft** [comment]
FYI: https://github.com/microsoft/onnxruntime/pull/8446

**Comment by snnn** [comment]
Close this PR as Chi has proposed a new PR and finished this work.

**Pull request closed without merging**